feat: Add more robust config defaulting

cdoern opened the pull request:
Currently,  one cannot specify a partial config. This is a useful thing to do when you don't care about some/most fields in the cfg, just GPU related ones.

pydantic has a `@model_validator` which allows a pre-check to be run before erroring out for missing fields. What I did in this commit is for each class and nested sub-class in `ilab`
the user can now pass little to no entires and we will fill out the rest of the config for them using the defaults in `get_default_config`. This differs from the current behavior which only allows the `Optional` fields to be omitted.

This config format will be useful in the following scenario:

A user can specify a config like the following:

```
serve:
  backend: null
  chat_template: null
  host_port: 127.0.0.1:8000
  llama_cpp:
    gpu_layers: -1
    llm_family: ''
    max_ctx_size: 4096
  model_path: /home/cdoern/.cache/instructlab/models/merlinite-7b-lab-Q4_K_M.gguf
  vllm:
    llm_family: ''
    vllm_args: [--tensor-parallel-size, 8]

```

if all they care about is serving in this scenario. `ilab` previously would error and tell them to fill out the entire cfg yaml with all mandatory fields. This level of strict enforcing makes sense for the `Config` class in code but not in everyday practice when passing yaml files. Now, ilab will fill out the entire config with sane defaults.


<!-- Thank you for contributing to InstructLab! -->

<!-- STEPS TO FOLLOW:
  1. Add a description of the changes (frequently the same as the commit description)
  2. Enter the issue number next to "Resolves #" below (if there is no tracking issue resolved, **remove that section**)
  3. Follow the steps in the checklist below, starting with the **Commit Message Formatting**.
-->

<!-- Uncomment this section with the issue number if an issue is being resolved
**Issue resolved by this Pull Request:**
Resolves #
--->

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the
  [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [ ] [Changelog](https://github.com/instructlab/instructlab/blob/main/CHANGELOG.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

RobotSail commented:
Nice, this worked perfectly on my local box to run training without the lora args.

booxter reviewed (commented):
I haven't reviewed in depth yet, because I have reservations about the very approach taken here. This feels like a kludge for a specific use case.

I think we should get away from the assumption that a config file is a single thing. Instead, the following would seem more flexible and would cover different use cases (distro level configs; system admin configs; users with multiple configurations; ad-hoc adjusting just a narrow set of configuration options for a particular target):

1. cascaded config file definitions (`ilab` being able to pick up "default" values - in order - first from "distro level" config files, then from "system level" config files, then from "user level" config files; then from additional "CLI option" config files; and finally, default to values embedded in the python code in `configuration.py` schemas.

All of the config files would then be merged / melded (honoring load priority) into a final configuration.

2. Eventually, add support for `conf.d` directories that would contain multiple config snippets, that would then be merged into a final configuration in runtime (with load order defined by e.g. lexicographical order of file names.)

---

I'd rather the project defines a flexible config loading mechanism and implement it once and for all than introduce kludges for specific use cases. We already have the "golden config" feature that will require reconciliation with a longer term design. What do you think?

With that in mind, does this really have to merge in `0.18.0`? Let me know.

cdoern commented:
@booxter I don't quite get what you mean by:
>  This feels like a kludge for a specific use case.

This PR simplifies (not complicates) the config usage. if a user wants to specify a config w/ just a train config, or just a generate config, `ilab` will take care of populating the rest.

This removes many situations we have run into while testing with extremely strict and IMO unrealistic expectations for users who do not have the full context of every entry in our config. 

I agree our config story should be revisited after 0.18.0 but I dont think redesigning our config and this PR are mutually exclusive.

Also this PR and the downstream cfg it introduces is a 0.18.0 requirement

cdoern commented on src/instructlab/configuration.py:
```diff
@@ -586,13 +733,72 @@ def ensure_storage_directories_exist():
+            deepspeed_cpu_offload_optimizer_ratio=1,
+            deepspeed_cpu_offload_optimizer_pin_memory=False,
+            is_padding_free=True,
+            model_path="~/.cache/instructlab/models/instructlab/granite-7b-starter",
```
this config is for linux primarily in 0.18.0, there is a downstream issue to track this and the downstream issue is a blocker

booxter replied:
But you propose it in upstream. It should reconcile with what upstream project supports.

---

Let me try to understand how you envision this used. So one would 1) call `ilab init` to produce both "default" and "internal" (what does the name mean?) config files, with different contents. 2) user would then `ilab ... --config-file .../internal.yaml` to use your custom configuration options. Correct?

Can't your distribution just include the "internal" config file in their packaging and instruct the user to pass `--config-file` pointing to it? (Perhaps giving the user a one-liner wrapper that would do it for them?)

This "internal config" seems rather opinionated. I feel these opinions do not belong to upstream. I think downstream should already be free to choose their "golden config" and target specific configuration files, if needed. What am I missing?

booxter commented:
@cdoern I think maybe you could give some details as to how this feature would be used. For what I see, you just dump an opinionated config file to `internal.yaml`. How is it then loaded? With `--config-file`? If so, how one passes "just a train config"?

cdoern commented:
@booxter I will update the description of this PR with some detailed examples to highlight the usage

nathan-weinberg commented on src/instructlab/configuration.py:
```diff
@@ -515,6 +708,20 @@ def read_config(
         raise ConfigException(msg) from exc
 
 
+# iter_nested_dict takes the config dictionary, iters over it via recursion, and each time it finds a relatve
```
```suggestion
# iter_nested_dict takes the config dictionary, iters over it via recursion, and each time it finds a relative
```

nathan-weinberg commented on src/instructlab/configuration.py:
```diff
@@ -515,6 +708,20 @@ def read_config(
 
 
+# iter_nested_dict takes the config dictionary, iters over it via recursion, and each time it finds a relatve
+# path, it replaces it with an absolute one. Recusrion is the easiest solution here to account for future cfg changes.
```
```suggestion
# path, it replaces it with an absolute one. Recursion is the easiest solution here to account for future cfg changes.
```

nathan-weinberg commented on src/instructlab/configuration.py:
```diff
@@ -586,13 +793,209 @@ def ensure_storage_directories_exist():
     }
 
+    internal_config = Config(
+        chat=_chat(model="~/.cache/instructlab/models/granite-7b-redhat-lab"),
+        serve=_serve(model_path="~/.cache/instructlab/models/granite-7b-redhat-lab"),
+        generate=_generate(
+            model="~/.cache/instructlab/models/mixtral-8x7b-v0-1",
+            gpus=4,
+            teacher=_serve(
+                backend="vllm",
+                model_path="~/.cache/instructlab/models/mixtral-8x7b-v0-1",
```
Is it alright that we are using `~` here? Will that work on non-Linux platforms?

booxter replied:
It won't. I pointed it out elsewhere.

JamesKunstle replied:
it works on mac / linux- those are all we're targeting right?

booxter replied:
AFAIU macosx does not use XDG paths. It puts cached models under Application Support directory. (I think there's an issue somewhere to understand if this should be changed to ~/.cache though.)

booxter replied:
(I also don't know if we explicitly target **only** *nix platforms. In theory, this is just python code that probably runs on win32 etc.? **I don't know if we care though**.)

booxter replied:
Here's the macosx bug: https://github.com/instructlab/instructlab/issues/1747

cdoern replied:
fixed

booxter commented on src/instructlab/configuration.py:
```diff
@@ -586,13 +793,209 @@ def ensure_storage_directories_exist():
+        with open(DEFAULTS.INTERNAL_CONFIG_FILE, "w", encoding="utf-8") as outfile:
+            yaml.dump(get_dict(internal_config), outfile)
+
+    if not os.path.isfile(DEFAULTS.TRAIN_A100_H100_X8_PROFILE):
```
Seems very repetitive. Probably a `for` loop would be neater.

cdoern replied:
good point, I will do that

booxter commented:
@cdoern I still don't understand how this is supposed to be used, and why shipping a custom config with whatever the settings a distributor prefers would not work for your needs. Could you please show a shell session output demoing the new feature? How is it different from passing `--config-file` pointing to a prepared `.yaml` file?

cdoern commented on src/instructlab/configuration.py:
```diff
@@ -515,6 +708,20 @@ def read_config(
+        if isinstance(value, dict):
+            yield from iter_nested_dict(value)
+        else:
+            if isinstance(value, str) is str and (
```
oops, sorry this is leftover from when I had `type(...)`

cdoern commented:
@booxter here are a few reasons I chose to do it this way:



all of the things we would be providing solely downstream are:

-  ilab specific files, so they inherently should live upstream
-  subject to change so rendering them this way allows for the most fluidity (not as hardcoded files)
- should live in persistent storage, the only way to render things into there currently is this approach
- able to be used by community members if they have a similar-ish amount of vram

That being said shipping these only downstream seems like an unnecessary moat that would turn away possible community and research interaction

cdoern commented:
no one is going to know how to really take advantage of all of our HW enablement if we hide it

booxter commented:
I admire the willingness to share profiles and other configs with community. I think we can introduce a special space for these. (Call a new directory `contribs` or `examples` and put stuff there, as `.yaml` files.)

I think this PR solves three separate issues: 1) default behavior when some settings are omitted - `ilab` should use in-code defaults instead of failing; 2) `internal.yaml` with red hat (?) specific settings; 3) a set of training profiles with red hat (?) specific settings. (Or at least opinionated settings.)

I think we should try to look at each of these on their own (in separate PRs). I have doubts that (2) and (3) belong to upstream; or at least that they belong to the production *code* (looks like the same could be achieved with `.yaml` files?) On the other hand, I think (1) is very valuable and is aligned with the overall configuration strategy for the future. It would be nice to land (1) ASAP.

RobotSail commented on src/instructlab/configuration.py:
```diff
@@ -586,13 +793,209 @@ def ensure_storage_directories_exist():
+            ckpt_output_dir="~/.local/share/instructlab/checkpoints",
+            data_output_dir="~/.local/share/instructlab/internal",
+            data_path="~/.local/share/instructlab/datasets",
+            nproc_per_node=4,
```
Should the internal config be using 4 GPUs by default?

RobotSail commented on src/instructlab/configuration.py:
```diff
@@ -515,6 +708,20 @@ def read_config(
 
+# iter_nested_dict takes the config dictionary, iters over it via recursion, and each time it finds a relatve
+# path, it replaces it with an absolute one. Recusrion is the easiest solution here to account for future cfg changes.
+def iter_nested_dict(config):
```
I recommmend type-hinting here since this semems like an important function and therefore should have a clear contract to help prevent errors.

JamesKunstle commented on src/instructlab/configuration.py:
```diff
@@ -300,6 +334,17 @@ def api_base(self):
+    ) -> typing.Dict[str, typing.Any]:
+        defaults = {
+            "model_path": DEFAULTS.DEFAULT_MODEL,
+            "llama_cpp": {"gpu_layers": -1, "max_ctx_size": 4096, "llm_family": ""},
```
could these hard-coded default values come from a config file instead of in the code?

rhatdan replied:
Hard coded default values should always come from code.  What happens if the default config files do not exist?

nathan-weinberg replied:
I wonder if we should have _all_ of defaults stored in the form of `DEFAULT.<name>` - e.g. in this case you could have `DEFAULTS.LLAMA_CPP_CONFIG` or something of the like. That way all the actual values are separated from the filler functions instead of a split. @cdoern wdyt?

cdoern replied:
I honestly think having the defaults in the function makes sense for readability purposes given that if they were extracted out, it would be another step in understanding what this code does @nathan-weinberg is that alright?

cdoern replied:
also, most of the `DEFAULTS.` things are variable paths that change based off of persistent storage. I think we should limit those to highly system dependent defaults, not constants like `-1` and `4096` etc

nathan-weinberg replied:
The readability point I feel kinda falls apart considering that you can only read _some_ of the values in the function and not others. If all the values were there that'd be one thing, but that's not the case.

That being said this is more of a design point, so I won't block on this.

JamesKunstle commented on src/instructlab/configuration.py:
```diff
@@ -390,6 +501,30 @@ class _train(BaseModel):
+            "data_path": DEFAULTS.DATASETS_DIR,
+            "ckpt_output_dir": DEFAULTS.CHECKPOINTS_DIR,
+            "data_output_dir": DEFAULTS.INTERNAL_DIR,
+            "max_seq_len": 4096,
```
same thing as I said before- I think these baseline defaults should come from an internal config file rather than being hard-coded into the config.

rhatdan replied:
Disagree.  The code needs to be usable without having any config files.

JamesKunstle commented on src/instructlab/configuration.py:
```diff
@@ -412,6 +547,63 @@ class Config(BaseModel):
+    defaults: dict, values: typing.Dict[str, typing.Any]
+) -> typing.Dict[str, typing.Any]:
+    for key, value in defaults.items():
+        if values.get(key) is None:
```
"values" vs. "value" collision is a little tricky to read. could be `set_values` instead

booxter replied:
Wondering if maybe instead of validating that it's `null`, we should validate that it's present. A user may want to override a default non-`null` value with a `null` - is it never the case?

JamesKunstle commented on src/instructlab/configuration.py:
```diff
@@ -586,13 +793,209 @@ def ensure_storage_directories_exist():
+    )
+
     # create exper_args file for users to see/edit
     if not os.path.isfile(DEFAULTS.TRAIN_ADDITIONAL_OPTIONS_FILE):
```
Is there any way we can avoid storing these large config blocks in the configuration code and use a set of config files instead? a middle-ground is having Python files for each of these config objects if moving some .yaml files around is a problem, but having data stored in code is kinda kludgey.

cdoern commented:
After conversations yesterday this PR is going to be split up into two.

This PR will add support for the partial config and will introduce a minimal "power user" config with upstream models

The other PR will introduce the training profiles _also_ with upstream models

These will be used as the basis for other configs to be shipped elsewhere

booxter commented on src/instructlab/config/init.py:
```diff
@@ -109,7 +109,6 @@ def init(
     cfg.serve.model_path = model_path
     cfg.generate.taxonomy_path = taxonomy_path
     cfg.generate.taxonomy_base = taxonomy_base
-    cfg.evaluate.model = model_path
```
is it intentional?

cdoern replied:
yes, it does not make sense for eval to have a default given that you evaluate the model given by training

booxter replied:
OK. This goes in the other direction from the other patch [adding](https://github.com/instructlab/instructlab/pull/1897) teacher model path though, so the two changes will have to reconcile somehow. I understand there are opinions about `--model-path` option (e.g. that it should be [removed](https://github.com/instructlab/instructlab/issues/1976).) but is it something this PR **has** to touch now?

booxter commented on src/instructlab/configuration.py:
```diff
@@ -586,13 +769,78 @@ def ensure_storage_directories_exist():
         ) as outfile:
             yaml.dump(additional_args_and_defaults, outfile)
 
+    if not os.path.isfile(DEFAULTS.INTERNAL_CONFIG_FILE):
```
I thought the plan was to not generate these in-code?..

cdoern replied:
no, this is still the plan as of the last conversation I believe

cdoern commented:
@booxter there might have been some miscommunication here. I believe the train profiles and internal config are still to be generated in this manner. I am unsure why the in code generating is a blocker here, we currently do this for other files stored in the persistent storage

cdoern commented:
this needs to merge before #1959 due to the new cfg defaulting in this PR

booxter commented:
I don't think **generating** these files in-code was a plan, as per last communication.

I still don't understand how a user is supposed to use this config. `--config-file <path-to-internal.yaml>`? Then it's no better than putting it as an example somewhere to use as a golden config, or through `--config-file`. I don't see why this has to be generated for anyone who may not care about red hat `internal.yaml` config.

 But I'd like @n1hility to chime in, because maybe I am delusional?..

(As a side note, I think this PR should have been split into several parts, where one distinct part would deal with the defaulting behavior. Generation of files / profiles has nothing to do with it. Unrelated topics should not be merged like this in a single PR. Mangling these into a single PR slows down the merge.)

cdoern commented:
one of the primary uses of this config is via `ILAB_GLOBAL_CONFIG` and the eventual support of a global system level ilab config.

I don't think splitting this up any more will have any benefits given the time it will take to do that and the timeline we are operating on. The config defaulting I added in this PR is tied to the internal cfg since it is necessary to fill out the rest of the defaults. 

As per last communication I believe @n1hility and @jaideepr97 did not have issues with the way I was generating these files.

booxter commented:
> one of the primary uses of this config is via `ILAB_GLOBAL_CONFIG`

OK, can you **please** elaborate on this? I've asked before to share details about how you think this new `internal.yaml` is supposed to be used. I haven't got an answer so far.

I have this idea of how you maybe expect this to be used, but it makes little sense. I was hoping that you could show where I am wrong.

Here is what I was thinking:

1. User calls `ilab config init` -> this results in `config.yaml` **and** `internal.yaml` dumped to disk.
2. User then picks the `internal` config file with `--config-file` (here, I would love to understand two things - first, when do they pick this config file and not the default one? how does a user decide to do it?; and second, what does "internal" in the name of the file mean? "internal" to what? To Red Hat? To `ilab`?)

From the above, I still don't see where `ILAB_GLOBAL_CONFIG` is in play though. Can you please try to explain it to me? Perhaps not just to me, but to other reviewers and users too - maybe this warrants a CHANGELOG update? I can't be the only one in the world confused about what this PR is trying to address, can I? :)

mairin commented:
Hey, wondering if this is really a 0.18.0 blocker? Providing a partial config is a convenient but not necessary, right? What would the consequence be to 0.18.0 if this didn't make it?

rhatdan commented:
I am sadly out of the look here, but I think this should follow the standard set for container tools.  Config defaults needs to be hard backed into the code, so if config files do  not exist, then these hard-coded configs are used.

Then use then as @booxter stated, the tool should look for Distro config in /usr/share/instructlab or ilab for the first conf,  This conf should be in the upstream repo with all of the hard coded defaults documented and commented out.  If a distro like RHEL wants to modify them, then it should.  The next step is /etc/instructlab for admin changes, no need to specify all defaults in any of these configs, if a field is not specified then use the next up in the hierarchy for default.

Next step is config in the users homedir, next step is overrides in ENV Variables if used, and finally at the command line.

If you support drop in config files in /etc/instuctlab.d Then these take precedence of /etc/instructlab, similar ones in homedirs.

BTW I prefer TOML to YAML for human consumption.

cdoern commented:
@rhatdan do you think the way I have done it here where we render our defaults Into yaml files is alright? Ilab has begun to use XRG dirs so these will live in ~/.local/share/instructlab/internal/config.yaml. 

I did it this way rather that having filled out yaml that we cp over to .local for more flexibility and less duplication of files on the system for users to mess with. 


As for yaml vs toml I agree bit unfortunately ilab already is pretty far down the yaml rabbit hole

cdoern commented:
@mairin I have provided versions of these configurations to @n1hility so I don't think we are totally blocked. However, having upstream versions of the downstream configurations to overwrite I think is important. In my eyes this pr can be merged in the state it's in. The only leftover debate is code generated configurations or hardcoded files. This pr generates them via code. At the end of the day it's the same end result but I feel pretty strongly about not duplicating the files and copying them into XDG dirs, having hardcoded defaults in code we directly render into XDG seems cleaner? Open to suggestions though

booxter commented:
For the record, I think I've come around to keeping defaults in-code. There are good reasons for this.

---

Probably what I still miss is how the alternative config file rendered **on disc** is supposed to be used, and when. Whether it's helpful to all users or a subset of them. Etc.

I see the description says it's for "4+ GPU" machines. Is it applicable for 2 GPU machines? (with a modification to tensor size option)? 6 GPU machines? (looks like we hardcode to 4?) Perhaps some documentation / CHANGELOG update would help me understand how users are supposed to think about this new config file. (And how they are expected to use it `--config-file` or `ILAB_GLOBAL_CONFIG`? There are different ideas mentioned here.)

---

@mairin 

> I also saw discussion that it would be ok to land the defaulting part of this PR; I don't know if that would be a 4th component this PR would split into or what, since the discussion seemed to lead to not generating profiles in code?

I think this part is not listed in your 1-3 list. This is one of the major parts of this PR that allows a user to omit including some (or all?) config options in their `config.yaml` if they don't care to override. (The "internal config" part relies on it in that it can list only the adjustments relevant to "power user" scenario, instead of including the complete yaml. But it could as well be implemented without "defaulting" change - by dumping **full** config to the `internal.yaml` file. It's just not as pretty.)

cdoern commented:
@booxter @rhatdan @mairin after discussing today, this PR is going to just be the more robust config defaulting.

If a user specifies a partial cfg, ilab no longer errors out on missing fields but fills out the rest instead

khaledsulayman commented on tests/test_config.py:
```diff
@@ -109,34 +109,25 @@ def test_default_config(self, cli_runner):  # pylint: disable=unused-argument
         self._assert_model_defaults(cfg)
 
-    def test_config_missing_required_field_groups(self, tmp_path_home):
+    def test_cfg_auto_fill(self, tmp_path_home):  # pylint: disable=unused-argument
```
are you sure the pylint rule is needed? Seems like you're using both args in the test

cdoern replied:
sorry, will remove

khaledsulayman commented on src/instructlab/configuration.py:
```diff
@@ -435,6 +543,63 @@ class Config(BaseModel):
+def finish_cfg_section(
+    defaults: dict, values: typing.Dict[str, typing.Any]
+) -> typing.Dict[str, typing.Any]:
+    for key, value in defaults.items():
+        if values.get(key) is None:
+            values[key] = value
```
```suggestion
    for key, val in defaults.items():
        if values.get(key) is None:
            values[key] = val
```

khaledsulayman reviewed (commented):
left some minor nits. Otherwise LGTM

RobotSail reviewed (approved):
LGTM

booxter commented on src/instructlab/configuration.py:
```diff
@@ -260,6 +260,13 @@ class _chat(BaseModel):
+    def fill_defaults(
+        cls, values: typing.Dict[str, typing.Any]
+    ) -> typing.Dict[str, typing.Any]:
+        defaults = {"model": DEFAULTS.DEFAULT_MODEL}
```
I probably miss something, but wouldn't this be achieved with pydantic by defining the `model` attribute as:

```
    model: str = Field(default_factory=lambda: "bla")
```

With that, I was able to remove the whole `chat:` section, only leaving `chat: {}`.

To go one step further, I added the default value for `_chat` attribute on `_config` class, and this allowed me to remove the remaining `chat: {}` section.

```
class Config(BaseModel):
    """Configuration for the InstructLab CLI."""

    # required fields
    chat: _chat = _chat()
...
```

What am I missing?

booxter replied:
PS: I think the proper thing to do for the `_chat` default may require a level of indirection through a `Field` too, like we do for any `@property` based `DEFAULTS.*`, e.g.:

```
chat: _chat = Field(default_factory=_chat)
```

(Note, this was **not** tested.)

Otherwise, test enviroment that overrides base directory for each test case may have these embedded structures unnecessarily cached.

booxter commented on src/instructlab/configuration.py:
```diff
@@ -311,6 +318,17 @@ def api_base(self):
+    ) -> typing.Dict[str, typing.Any]:
+        defaults = {
+            "model_path": DEFAULTS.DEFAULT_MODEL,
+            "llama_cpp": {"gpu_layers": -1, "max_ctx_size": 4096, "llm_family": ""},
```
This seems to repeat what is defined in the attribute defaults, e.g. for `llama_cpp`, the default used is already listed in line 302. The other listed fields have defaults too. What does this function (and other similar functions here) give that the `= <default_value> / Field(lambda: <default_value>)` wouldn't be able to achieve?

booxter reviewed (commented):
What am I missing? Don't `pydantic` defaults already work? (They seem to work for me here.)

cdoern commented:
@booxter , currently if one provides a cfg yaml (via the --config flag or `ILAB_GLOBAL_CONFIG`) if not complete, ilab errors. This allows users to give us a very minimal cfg which we will fill out the rest of. This is approaching from a yaml perspective not an in code class perspective.

booxter commented:
@cdoern I understand the intent of the patch, and I **violently agree** with it. The end state is being able to run `ilab` commands without a config file at all. (just from in-code defaults). 🙏

What I question is that maybe we don't need the `fill_defaults` etc. at all, we can probably rely on `pydantic` defaults. We already use them for a lot of fields, so we may just need to fill in the gaps.

To demonstrate what I mean, see the following PR - I will mark it as draft; it includes your patch and one more on top that achieves the same goal but without `fill_defaults` etc. Let me know what you think about it: https://github.com/instructlab/instructlab/pull/1985

booxter commented on src/instructlab/configuration.py:
```diff
@@ -339,25 +357,74 @@ class _generate(BaseModel):
+        cls, values: typing.Dict[str, typing.Any]
+    ) -> typing.Dict[str, typing.Any]:
+        defaults = {
+            "judge_model": DEFAULTS.JUDGE_MODEL_MT,
```
I think there's a bug here: the defaults dict doesn't match the class. Not that I think this code should stay, but in case we want to stick to this version of the implementation, this will have to be fixed.

cdoern replied:
fixed!

JamesKunstle commented on src/instructlab/configuration.py:
```diff
@@ -373,6 +440,23 @@ class _evaluate(BaseModel):
+                "output_dir": DEFAULTS.EVAL_DATA_DIR,
+                "max_workers": 16,
+            },
+            "mmlu": {"few_shots": 16, "batch_size": "auto"},
```
few_shot should *definitely* be 5, not 16

cdoern replied:
fixed

cdoern commented:
Thanks for listing the concerns out like this @booxter ! I will work on addressing those now

cdoern commented:
ok here is what I did

1. fixed the mtbenchbranch config defaults which were duplicated from mtbench
2. fixed the eval model path (added it back for now)
3. fixed the few shots (5 rather than 16)
4. removed the path expansion

cdoern commented on src/instructlab/configuration.py:
```diff
@@ -544,6 +710,20 @@ def read_config(
+            yield from iter_nested_dict(value)
+        else:
+            if isinstance(value, str) and (
+                os.path.isdir(value) or os.path.isfile(value)
```
removed this in the current version
